src/core/lifetime.d

```diff
@@ -2292,3 +2353,186 @@ void _d_delstruct(T)(ref T *p)
     assert(innerDtors == 2);
     assert(outerDtors == 1);
 }
+
+// issue 25552
+@safe unittest
+{
+    int i;
+    struct Nested
+    {
+        void foo() { ++i; }
+        ~this() {}
+    }
+
+    static struct NotNested
+    {
+        Nested n;
+    }
+
+    static struct Deep
+    {
+        NotNested nn;
+    }
+
+    static struct Deeper
+    {
+        NotNested[1] nn;
+    }
+
+    static assert(__traits(isZeroInit, Nested));
+    static assert(__traits(isZeroInit, NotNested));
+    static assert(__traits(isZeroInit, Deep));
+    static assert(__traits(isZeroInit, Deeper));
+
+    auto a = NotNested(Nested());
+    assert(a.n.tupleof[$-1]);
+    auto b = move(a);
+    assert(b.n.tupleof[$-1]);
+    assert(a.n.tupleof[$-1] is b.n.tupleof[$-1]);
+
+    auto c = Deep(NotNested(Nested()));
+    auto d = move(c);
+    assert(d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.tupleof[$-1] is d.nn.n.tupleof[$-1]);
+
+    auto e = Deeper([NotNested(Nested())]);
+    auto f = move(e);
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+}
+
+// issue 25552
+@safe unittest
+{
+    int i;
+    struct Nested
+    {
+        int[3] stuff = [0, 1, 2];
+        void foo() { ++i; }
+        ~this() {}
+    }
+
+    static struct NoAssign
+    {
+        int value;
+        @disable void opAssign(typeof(this));
+    }
+
+    static struct NotNested
+    {
+        int before = 42;
+        Nested n;
+        auto after = NoAssign(43);
+    }
+
+    static struct Deep
+    {
+        NotNested nn;
+    }
+
+    static struct Deeper
+    {
+        NotNested[1] nn;
+    }
+
+    static assert(!__traits(isZeroInit, Nested));
+    static assert(!__traits(isZeroInit, NotNested));
+    static assert(!__traits(isZeroInit, Deep));
+    static assert(!__traits(isZeroInit, Deeper));
+
+    auto a = NotNested(1, Nested([3, 4, 5]), NoAssign(2));
+    auto b = move(a);
+    assert(b.n.tupleof[$-1]);
+    assert(a.n.tupleof[$-1] is b.n.tupleof[$-1]);
+    assert(a.n.stuff == a.n.init.stuff);
+    assert(a.before == a.init.before);
+    assert(a.after == a.init.after);
+
+    auto c = Deep(NotNested(1, Nested([3, 4, 5]), NoAssign(2)));
+    auto d = move(c);
+    assert(d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.tupleof[$-1] is d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.stuff == c.nn.n.init.stuff);
+    assert(c.nn.before == c.nn.init.before);
+    assert(c.nn.after == c.nn.init.after);
+
+    auto e = Deeper([NotNested(1, Nested([3, 4, 5]), NoAssign(2))]);
+    auto f = move(e);
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.stuff == e.nn[0].n.init.stuff);
+    assert(e.nn[0].before == e.nn[0].init.before);
+    assert(e.nn[0].after == e.nn[0].init.after);
+}
+
+// wipes source after moving (zero init case)
+pragma(inline, true)
+private void wipe(T)(return scope ref T source) @trusted
```

Comment:
It might make sense to try to combine both wipe-versions into one, as the basic static-if structure is the same - maybe as a variadic template, with a single optional `const scope ref T initializer` param.

Reply:
Is there a benefit to doing that? Going to be the same code, just harder to read, and the same amount of symbols generated. I mean, I can do that, just not sure if it's worth doing.

Reply:
> Cuz readers of such code definitely won't :D

On the contrary, my guess is that it would improve readability by reducing duplication and number of lines. After all, the only difference is doing a memset-to-0 or memcpy in 3 lines, and optionally passing the initializer for the 2 recursive calls.

Reply:
But the `Args...`, the constraint, the `static if (Args.length)`... Anyway, let's just do it and see how it reads.